jax_distribution/evaluator.py

```diff
@@ -138,35 +125,212 @@ def evaluator_fn(trained_params: FrozenDict, rng: chex.PRNGKey) -> ExperimentOut
+        """Evaluator function."""
+
+        # Initialise environment states and timesteps.
+        n_devices = len(jax.devices())
+
+        eval_batch = config["num_eval_episodes"] // n_devices * eval_multiplier
+
+        rng, *env_rngs = jax.random.split(rng, eval_batch + 1)
+        env_states, timesteps = jax.vmap(env.reset)(
+            jnp.stack(env_rngs),
+        )
+        # Split rngs for each core.
+        rng, *step_rngs = jax.random.split(rng, eval_batch + 1)
+        # Add dimension to pmap over.
+        reshape_step_rngs = lambda x: x.reshape(eval_batch, -1)
+        step_rngs = reshape_step_rngs(jnp.stack(step_rngs))
+
+        # Initialise hidden state.
+        init_hstate = scanned_rnn.initialize_carry(eval_batch, 128)
+        init_hstate = jnp.expand_dims(init_hstate, axis=1)
+        init_hstate = jnp.expand_dims(init_hstate, axis=2)
+        init_hstate = jnp.tile(init_hstate, (1, config["num_agents"], 1))
+
+        if centralised_critic:
+            init_critic_hstate = scanned_rnn.initialize_carry(eval_batch, 128)
+            init_critic_hstate = jnp.expand_dims(init_critic_hstate, axis=1)
+            init_hstate = (init_hstate, init_critic_hstate)
+
+        # Initialise dones.
+        dones = jnp.zeros(
+            (
+                eval_batch,
+                config["num_agents"],
+            ),
+            dtype=bool,
+        )
+
+        eval_state = RNNEvalState(step_rngs, env_states, timesteps, dones, init_hstate)
+
+        eval_metrics = jax.vmap(eval_one_episode, in_axes=(None, 0), axis_name="eval_batch")(
+            trained_params, eval_state
+        )
+
+        return ExperimentOutput(
+            episodes_info=eval_metrics,
+        )
+
+    return evaluator_fn
+
+
 def evaluator_setup(
     eval_env: Environment,
     rng_e: chex.PRNGKey,
     network: Any,
     params: FrozenDict,
     config: Dict,
+    centralised_critic: bool = False,
+    rnn_net: bool = False,
+    scanned_rnn: nn.Module = None,
 ) -> Tuple[callable, callable, Tuple]:
     """Initialise evaluator_fn, network, optimiser, environment and states."""
     # Get available TPU cores.
     n_devices = len(jax.devices())
 
-    # Vmap it over number of agents.
-    vmapped_eval_network_apply_fn = jax.vmap(
-        network.apply,
-        in_axes=(None, 0),
-    )
-
-    evaluator = get_evaluator_fn(
-        eval_env,
-        vmapped_eval_network_apply_fn,
-        config,
-    )
-
-    absolute_metric_evaluator = get_evaluator_fn(
-        eval_env,
-        vmapped_eval_network_apply_fn,
-        config,
-        10,
-    )
+    # Vmap it over number of agents and create evaluator_fn.
+    if rnn_net:
+        if centralised_critic:
+            vmapped_eval_network_apply_fn = jax.vmap(
+                network.apply,
+                in_axes=(
+                    None,
+                    (1, None),
+                    (ObservationGlobalState(2, 2, None, 2), None),
```

Comment:
Howcome you don't need to do this if we're ff centralised?

Reply:
We already did the same for the mappo check this[ line](https://github.com/instadeepai/Mava/blob/feat/pure-jax-mava/jax_distribution/systems/ff_mappo_rware.py#L355)